lib/tickets/ticket_query_object.rb

```diff
@@ -2,61 +2,58 @@
 
 class TicketQueryObject
   def initialize(params)
-    @search = params[:search]
-    @what = params[:what]
+    @whats = {
+      by_email_or_username: params[:by_email_or_username],
+      in_subject: params[:in_subject],
+      in_content: params[:in_content],
+      by_course: params[:by_course]
+    }
+    @whats.delete_if { |_, value| value.nil? || value.empty? }
     @offset = params[:offset] || 0
     @limit = params[:limit] || 100
   end
 
   def search
-    case @what
-    when 'by_email_or_username'
-      search_by_username_or_by_email
-    when 'in_subject'
-      search_by_subject
-    when 'in_content'
-      search_by_content
-    when 'by_course'
-      search_by_course
+    query = tickets
+
+    @whats.each do |key, value|
+      case key
+      when :by_email_or_username
+        query = query.merge(search_by_username_or_by_email(value))
```

Comment:
This is an interesting strategy, using `merge` to combine the query parameters. Nice!